source/agora/test/QuorumShuffle.d

```diff
@@ -0,0 +1,203 @@
+
+    Tests regular quorum shuffling behavior.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.QuorumShuffle;
+
+version (unittest):
+
+import agora.common.Serializer;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.Transaction;
+import agora.test.Base;
+
+/// With a validator cycle of 10 and shuffle cycle of 3 we should expect
+/// quorums being shuffled at these block heights:
+/// 3, 6, 9, 10 (enrollment change), 13 (next shuffle cycle)
+unittest
+{
+    import agora.common.Types;
+    TestConf conf = {
+        validators : 6,
+        max_listeners : 7,
+        extra_blocks : 2,  // 1 short of shuffle cycle
+        validator_cycle : 10,
+        max_quorum_nodes : 4,  // makes it easier to test shuffle cycling
+        quorum_shuffle_interval : 3
+    };
+    auto network = makeTestNetwork(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getBlockHeight() == 2, 5.seconds,
+            format("Node %s has block height %s. Expected: %s",
+                idx, node.getBlockHeight(), 2)));
+
+    const keys = WK.Keys.byRange.map!(kp => kp.address).take(6).array;
+
+    // 0 because the shuffle occured at genesis, not at block height 2
+    const quorums_1 = nodes[0].getExpectedQuorums(keys, Height(0));
+    // writeln(quorums_1);  // uncomment to check
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getQuorumConfig() == quorums_1[idx], 5.seconds,
+            format("Node %s has quorum config %s. Expected: %s",
+                idx, node.getQuorumConfig(), quorums_1[idx])));
+
+    // check that the preimages were revealed before we trigger block creation
+    // note: this is a workaround. A block should not be accepted if there
+    // are no preimages for this height - however currently block signatures
+    // are not implemented yet (#365)
+    const enrollments = network.blocks[0].header.enrollments;
+    void checkDistance (uint distance)
+    {
+        nodes.each!(node =>
+            enrollments.each!(enr =>
+                retryFor(node.getPreimage(enr.utxo_key).distance >= distance,
+                    5.seconds)));
+    }
+
+    checkDistance(3);
+
+    Transaction[] txs = network.blocks[$ - 1].txs.serializeFull
+        .deserializeFull!(Transaction[]);
+
+    void makeBlock (ulong height)
+    {
+        txs = txs.filter!(tx => tx.type == TxType.Payment)
+            .map!(tx => iota(tx.outputs.length)
+                .map!(idx => TxBuilder(tx, cast(uint)idx)))
+            .joiner().takeExactly(8)
+            .map!(txb => txb.refund(WK.Keys.Genesis.address).sign()).array;
```

Comment:
- Use `spendable()`;
- Don't use `Genesis`

Reply:
What do you mean by "don't use Genesis". The special Genesis address?

Reply:
> The special Genesis address?

Yes

Reply:
`spendable` requires us to do a lot of back and forth btw - I can see this is going to become another code segment that is copy-pasted everywhere:


```D
const(Block)[] blocks = [network.blocks[$ - 1]];

void makeBlock (ulong height)
{
    auto txs = blocks[$ - 1].spendable.takeExactly(8)
        .map!(txb => txb.refund(WK.Keys.Genesis.address).sign()).array;
    txs.each!(tx => nodes[0].putTransaction(tx));
    nodes.enumerate.each!((idx, node) =>
        retryFor(node.getBlockHeight() == height, 5.seconds,
            format("Node %s has block height %s. Expected: %s",
                idx, node.getBlockHeight(), height)));
    retryFor((blocks = nodes[0].getBlocksFrom(height, 1)).length == 1, 5.seconds,
            format("Node 0 getBlocksFrom(%s, 1) failed", height));
}
```

Reply:
1) Do we need `takeExactly` since we're transitioning away from tx-count-based block creation ?
2) Why do you set the refund address ? `sign()` is enough.
3) Since the transactions are  not  used as an array anywhere, no need to `array` then `each`, just `each` directly.

In short:
```D
blocks[$ - 1].spendable.each!(txb => nodes[0].putTransaction(txb.sign()));
```

Reply:
Oh sorry I didn't realize `sign` does this already.

> Do we need takeExactly since we're transitioning away from tx-count-based block creation ?

We'll we're still transitioning. :) 

Once we transition we can get rid of that, sure.

Reply:
> Oh sorry I didn't realize `sign` does this already.

Yes, the only requirement for `TxBuilder` is a refund address. If you provide it with a single output, it'll use the beneficiary as refund address, and if you provide it with a transaction, it'll use the first output's beneficiary by default.